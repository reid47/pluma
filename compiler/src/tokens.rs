// TODO: use this to only derive in tests
// #[cfg_attr(test, derive(Debug, PartialEq))]
#[derive(Debug, PartialEq)]
pub enum Token<'a> {
  Arrow { start: usize, end: usize },
  BinaryDigits { start: usize, end: usize, value: &'a [u8] },
  Colon { start: usize, end: usize },
  ColonEquals { start: usize, end: usize },
  Comma { start: usize, end: usize },
  Comment { start: usize, end: usize, value: &'a [u8] },
  DecimalDigits { start: usize, end: usize, value: &'a [u8] },
  Dot { start: usize, end: usize },
  DoubleArrow { start: usize, end: usize },
  DoubleColon { start: usize, end: usize },
  Equals { start: usize, end: usize },
  HexDigits { start: usize, end: usize, value: &'a [u8] },
  Identifier { start: usize, end: usize, value: &'a [u8] },
  InterpolationEnd { start: usize, end: usize },
  InterpolationStart { start: usize, end: usize },
  LeftBrace { start: usize, end: usize },
  LeftBracket { start: usize, end: usize },
  LeftParen { start: usize, end: usize },
  LineBreak { start: usize, end: usize },
  Minus { start: usize, end: usize },
  OctalDigits { start: usize, end: usize, value: &'a [u8] },
  Pipe { start: usize, end: usize },
  Plus { start: usize, end: usize },
  RightBrace { start: usize, end: usize },
  RightBracket { start: usize, end: usize },
  RightParen { start: usize, end: usize },
  StringLiteral { start: usize, end: usize, value: &'a [u8] },
  Unexpected { start: usize, end: usize },
}

pub fn get_token_location(token: &Token) -> (usize, usize) {
  match token {
    Token::Arrow { start, end } => (*start, *end),
    Token::BinaryDigits { start, end, .. } => (*start, *end),
    Token::Colon { start, end } => (*start, *end),
    Token::ColonEquals { start, end } => (*start, *end),
    Token::Comma { start, end } => (*start, *end),
    Token::Comment { start, end, .. } => (*start, *end),
    Token::DecimalDigits { start, end, .. } => (*start, *end),
    Token::Dot { start, end } => (*start, *end),
    Token::DoubleArrow { start, end } => (*start, *end),
    Token::DoubleColon { start, end } => (*start, *end),
    Token::Equals { start, end } => (*start, *end),
    Token::HexDigits { start, end, .. } => (*start, *end),
    Token::Identifier { start, end, .. } => (*start, *end),
    Token::InterpolationEnd { start, end } => (*start, *end),
    Token::InterpolationStart { start, end } => (*start, *end),
    Token::LeftBrace { start, end } => (*start, *end),
    Token::LeftBracket { start, end } => (*start, *end),
    Token::LeftParen { start, end } => (*start, *end),
    Token::LineBreak { start, end } => (*start, *end),
    Token::Minus { start, end } => (*start, *end),
    Token::OctalDigits { start, end, .. } => (*start, *end),
    Token::Pipe { start, end, .. } => (*start, *end),
    Token::Plus { start, end, .. } => (*start, *end),
    Token::RightBrace { start, end } => (*start, *end),
    Token::RightBracket { start, end } => (*start, *end),
    Token::RightParen { start, end } => (*start, *end),
    Token::StringLiteral { start, end, .. } => (*start, *end),
    Token::Unexpected { start, end } => (*start, *end),
  }
}