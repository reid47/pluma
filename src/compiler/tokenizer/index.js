import tokenize from './tokenizer';
export default tokenize;
