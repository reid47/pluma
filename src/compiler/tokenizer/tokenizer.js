import {
  tokenTypes,
  symbols,
  symbolRegexes,
  reservedWords,
  reservedWordRegexes
} from '../constants';

import { TokenizerError } from '../errors';

const patterns = {
  newline: /^[\n]+/,
  otherWhitespace: /^[ \t\r]+/,
  booleanLiteral: /^(True|False)/,
  decimalNumericLiteral: /^-?[\d]+\.?[\d]*/,
  hexNumericLiteral: /^-?0x[\da-f]+/i,
  octalNumericLiteral: /^-?0o[0-7]+/i,
  binaryNumericLiteral: /^-?0b[01]+/i,
  specialNumericLiteral: /^(NaN|Infinity)/,
  identifier: /^[a-z$_][0-9a-z$_-]*/i,
  atIdentifier: /^@[a-z$_][0-9a-z$_-]*/i,
  dotIdentifier: /^\.[a-z$_][0-9a-z$_-]*/i,
  digits: /^[0-9]+/
};

function tokenize({ source }) {
  const length = source.length;
  const tokens = [];
  const interpolationStack = [];

  let line = 1;
  let column = 0;
  let index = 0;
  let inString = false;
  let stringStart = null;
  let inRegex = false;
  let regexStart = null;
  let match;
  let remaining;

  function fail(
    message,
    { lineNumber = line, columnStart = column, columnEnd = column, hint } = {}
  ) {
    throw new TokenizerError(
      message,
      source,
      lineNumber,
      columnStart,
      columnEnd,
      hint
    );
  }

  function pushToken(type, text, value) {
    tokens.push({
      type,
      value,
      lineStart: line,
      lineEnd: line,
      columnStart: column,
      columnEnd: column + text.length
    });
  }

  function advance(amount = 1) {
    index += amount;
    column += amount;
  }

  outer: while (index < length) {
    remaining = source.substring(index);

    if (inString) {
      if (remaining[0] === '\n') {
        index++;
        column = 0;
        line++;
        continue;
      }

      const endQuote = remaining[0] === '"';
      const startInterpolation = remaining[0] === '$' && remaining[1] === '{';

      if (endQuote || startInterpolation) {
        const { charIndex, lineStart, columnStart } = stringStart;
        stringStart = null;
        inString = false;

        tokens.push({
          type: tokenTypes.STRING,
          value: source.substring(charIndex, index),
          lineStart,
          lineEnd: line,
          columnStart,
          columnEnd: endQuote ? column + 1 : column
        });

        if (endQuote) {
          advance();
        } else {
          interpolationStack.push(true);
        }

        continue;
      }

      advance();
      continue;
    }

    if (
      inRegex &&
      (remaining[0] !== '/' ||
        (remaining[0] === '/' && source[index - 1] === '\\'))
    ) {
      advance();
      continue;
    }

    if (remaining[0] === '"') {
      inString = true;
      stringStart = {
        lineStart: line,
        columnStart: column,
        charIndex: index + 1
      };

      advance();
      continue;
    }

    if (interpolationStack.length && remaining[0] === '}') {
      inString = true;
      stringStart = {
        lineStart: line,
        columnStart: column + 1,
        charIndex: index + 1
      };

      interpolationStack.pop();

      pushToken(tokenTypes.SYMBOL, '}', '}');

      advance();
      continue;
    }

    if ((match = remaining.match(patterns.newline))) {
      index += match[0].length;
      line += match[0].length;
      column = 0;
      continue;
    }

    if ((match = remaining.match(patterns.otherWhitespace))) {
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.booleanLiteral))) {
      pushToken(tokenTypes.BOOLEAN, match[0], match[0] === 'True');

      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.hexNumericLiteral))) {
      const isNegative = match[0][0] === '-';
      const parsed = parseInt(match[0].substring(isNegative ? 3 : 2), 16);

      pushToken(tokenTypes.NUMBER, match[0], parsed * (isNegative ? -1 : 1));

      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.octalNumericLiteral))) {
      const isNegative = match[0][0] === '-';
      const parsed = parseInt(match[0].substring(isNegative ? 3 : 2), 8);

      pushToken(tokenTypes.NUMBER, match[0], parsed * (isNegative ? -1 : 1));

      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.binaryNumericLiteral))) {
      const isNegative = match[0][0] === '-';
      const parsed = parseInt(match[0].substring(isNegative ? 3 : 2), 2);

      pushToken(tokenTypes.NUMBER, match[0], parsed * (isNegative ? -1 : 1));

      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.decimalNumericLiteral))) {
      pushToken(tokenTypes.NUMBER, match[0], parseFloat(match[0]));

      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.specialNumericLiteral))) {
      pushToken(tokenTypes.NUMBER, match[0], +match[0]);

      advance(match[0].length);
      continue;
    }

    if (remaining[0] === '/') {
      if (!inRegex) {
        inRegex = true;
        regexStart = {
          lineStart: line,
          columnStart: column,
          charIndex: index + 1
        };

        advance();
      } else {
        advance();
        const { lineStart, columnStart, charIndex } = regexStart;
        regexStart = null;
        inRegex = false;

        const regexString = source.substring(charIndex, index - 1);

        let regexFlags;
        if ((match = source.substring(index).match(/^[gimuy]+/))) {
          regexFlags = match[0];

          advance(match[0].length);
        }

        let value;
        try {
          value = new RegExp(regexString, regexFlags);
        } catch (err) {
          let hint;
          if (err.message.indexOf('Unterminated group') > -1) {
            hint =
              'It looks like you may be missing a closing ")" for a group.';
          } else if (err.message.indexOf('Unmatched ")"')) {
            hint =
              'It looks like you have a closing ")" without an opening "(".';
          }

          fail('Invalid regular expression', {
            lineStart,
            columnStart,
            columnEnd: column,
            hint
          });
        }

        tokens.push({
          type: tokenTypes.REGEX,
          value,
          lineStart,
          lineEnd: line,
          columnStart,
          columnEnd: column
        });
      }
      continue;
    }

    for (let w = 0; w < reservedWords.length; w++) {
      if (reservedWordRegexes[w].test(remaining)) {
        pushToken(tokenTypes.KEYWORD, reservedWords[w], reservedWords[w]);

        advance(reservedWords[w].length);
        continue outer;
      }
    }

    if ((match = remaining.match(patterns.dotIdentifier))) {
      pushToken(tokenTypes.DOT_IDENTIFIER, match[0], match[0].substring(1));

      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.atIdentifier))) {
      pushToken(tokenTypes.AT_IDENTIFIER, match[0], match[0].substring(1));

      advance(match[0].length);
      continue;
    }

    for (let s = 0; s < symbols.length; s++) {
      if (symbolRegexes[s].test(remaining)) {
        pushToken(tokenTypes.SYMBOL, symbols[s], symbols[s]);

        advance(symbols[s].length);
        continue outer;
      }
    }

    if ((match = remaining.match(patterns.identifier))) {
      pushToken(tokenTypes.IDENTIFIER, match[0], match[0]);

      advance(match[0].length);
      continue;
    }

    fail(`Unrecognized character '${remaining[0]}'`);
  }

  return tokens;
}

export default tokenize;
