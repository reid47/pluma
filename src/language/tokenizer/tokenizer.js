import {
  tokenTypes,
  symbols,
  symbolRegexes,
  reservedWords,
  reservedWordRegexes
} from '../constants';

import { TokenizerError } from '../errors';

const patterns = {
  newline: /^[\n]+/,
  otherWhitespace: /^[ \t\r]+/,
  booleanLiteral: /^(True|False)/,
  decimalNumericLiteral: /^[\d]+\.?[\d]*/,
  hexNumericLiteral: /^0x[\da-f]+/i,
  octalNumericLiteral: /^0o[0-7]+/i,
  binaryNumericLiteral: /^0b[01]+/i,
  specialNumericLiteral: /^(NaN|Infinity)/,
  identifier: /^[a-z$_][0-9a-z$_]*/i,
  digits: /^[0-9]+/
};

function tokenize({ source }) {
  const length = source.length;
  const tokens = [];
  const interpolationStack = [];
  let line = 1;
  let column = 0;
  let i = 0;
  let inString = false;
  let stringStart = null;
  let inRegex = false;
  let regexStart = null;
  let match;
  let remaining;

  function fail(
    message,
    { lineNumber = line, columnStart = column, columnEnd = column, hint } = {}
  ) {
    throw new TokenizerError(
      message,
      source,
      lineNumber,
      columnStart,
      columnEnd,
      hint
    );
  }

  function pushToken(type, text, value) {
    tokens.push({
      type,
      value,
      lineStart: line,
      lineEnd: line,
      columnStart: column,
      columnEnd: column + text.length
    });
  }

  function advance(amount = 1) {
    i += amount;
    column += amount;
  }

  outer: while (i < length) {
    remaining = source.substring(i);

    if (inString) {
      if (remaining[0] === '\n') {
        i++;
        column = 0;
        line++;
        continue;
      }

      const endQuote = remaining[0] === '"';
      const startInterpolation = remaining[0] === '$' && remaining[1] === '{';
      if (endQuote || startInterpolation) {
        const { charIndex, lineStart, columnStart } = stringStart;
        stringStart = null;

        tokens.push({
          type: tokenTypes.STRING,
          value: source.substring(charIndex, i),
          lineStart,
          lineEnd: line,
          columnStart,
          columnEnd: endQuote ? column + 1 : column
        });

        inString = false;
        if (endQuote) {
          advance();
        } else {
          interpolationStack.push(true);
        }

        continue;
      } else {
        advance();
        continue;
      }
    }

    if (
      inRegex &&
      (remaining[0] !== '/' || (remaining[0] === '/' && source[i - 1] === '\\'))
    ) {
      advance();
      continue;
    }

    if (remaining[0] === '"') {
      inString = true;
      stringStart = { lineStart: line, columnStart: column, charIndex: i + 1 };
      advance();
      continue;
    } else if (interpolationStack.length && remaining[0] === '}') {
      inString = true;
      stringStart = {
        lineStart: line,
        columnStart: column + 1,
        charIndex: i + 1
      };
      interpolationStack.pop();
      pushToken(tokenTypes.SYMBOL, '}', '}');
      advance();
      continue;
    }

    if ((match = remaining.match(patterns.newline))) {
      i += match[0].length;
      column = 0;
      line += match[0].length;
      continue;
    }

    if ((match = remaining.match(patterns.otherWhitespace))) {
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.booleanLiteral))) {
      pushToken(tokenTypes.BOOLEAN, match[0], match[0] === 'True');
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.hexNumericLiteral))) {
      pushToken(
        tokenTypes.NUMBER,
        match[0],
        parseInt(match[0].substring(2), 16)
      );
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.octalNumericLiteral))) {
      pushToken(
        tokenTypes.NUMBER,
        match[0],
        parseInt(match[0].substring(2), 8)
      );
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.binaryNumericLiteral))) {
      pushToken(
        tokenTypes.NUMBER,
        match[0],
        parseInt(match[0].substring(2), 2)
      );
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.decimalNumericLiteral))) {
      pushToken(tokenTypes.NUMBER, match[0], parseFloat(match[0]));
      advance(match[0].length);
      continue;
    }

    if ((match = remaining.match(patterns.specialNumericLiteral))) {
      pushToken(tokenTypes.NUMBER, match[0], +match[0]);
      advance(match[0].length);
      continue;
    }

    if (remaining[0] === '/') {
      if (!inRegex) {
        inRegex = true;
        regexStart = {
          lineStart: line,
          columnStart: column,
          charIndex: i + 1
        };
        advance();
      } else {
        advance();
        inRegex = false;
        const { lineStart, columnStart, charIndex } = regexStart;
        regexStart = null;

        const regexString = source.substring(charIndex, i - 1);

        let regexFlags;
        if ((match = source.substring(i).match(/^[gimuy]+/))) {
          regexFlags = match[0];
          advance(match[0].length);
        }

        let value;
        try {
          value = new RegExp(regexString, regexFlags);
        } catch (err) {
          let hint;
          if (err.message.indexOf('Unterminated group') > -1) {
            hint =
              'It looks like you may be missing a closing ")" for a group.';
          } else if (err.message.indexOf('Unmatched ")"')) {
            hint =
              'It looks like you have a closing ")" without an opening "(".';
          }

          fail('Invalid regular expression', {
            lineStart,
            columnStart,
            columnEnd: column,
            hint
          });
        }

        tokens.push({
          type: tokenTypes.REGEX,
          value,
          lineStart,
          lineEnd: line,
          columnStart,
          columnEnd: column
        });
      }
      continue;
    }

    if (remaining.lastIndexOf('null', 0) === 0) {
      pushToken(tokenTypes.NULL, 'null', null);
      advance(4);
      continue;
    }

    if (remaining.lastIndexOf('undefined', 0) === 0) {
      pushToken(tokenTypes.UNDEFINED, 'undefined', undefined);
      advance(9);
      continue;
    }

    for (let w = 0; w < reservedWords.length; w++) {
      if (reservedWordRegexes[w].test(remaining)) {
        pushToken(tokenTypes.KEYWORD, reservedWords[w], reservedWords[w]);
        advance(reservedWords[w].length);
        continue outer;
      }
    }

    for (let s = 0; s < symbols.length; s++) {
      if (symbolRegexes[s].test(remaining)) {
        pushToken(tokenTypes.SYMBOL, symbols[s], symbols[s]);
        advance(symbols[s].length);
        continue outer;
      }
    }

    if ((match = remaining.match(patterns.identifier))) {
      pushToken(tokenTypes.IDENTIFIER, match[0], match[0]);
      advance(match[0].length);
      continue;
    }

    fail(`Unrecognized character '${remaining[0]}'`);
  }

  return tokens;
}

export default tokenize;
