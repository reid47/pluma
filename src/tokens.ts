export type TokenKind =
  | 'Boolean'
  | 'Char'
  | 'Comment'
  | 'Identifier'
  | 'InterpolationEnd'
  | 'InterpolationStart'
  | 'Operator'
  | 'Number'
  | 'String'
  | '->'
  | '=>'
  | '::'
  | ':'
  | ':='
  | ','
  | '.'
  | '='
  | '{'
  | '}'
  | '['
  | ']'
  | '('
  | ')';

export interface Token {
  kind: TokenKind;
  value: string;
  lineStart: number;
  colStart: number;
  lineEnd: number;
  colEnd: number;
}
