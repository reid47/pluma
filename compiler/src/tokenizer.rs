use crate::tokens::{Token, Token::*};
use crate::tokenizer::{TokenizeResult::*, TokenizeError::*};
use std::collections::HashMap;

pub struct Tokenizer<'a> {
  source: &'a Vec<u8>,
  source_length: usize,
}

fn is_identifier_start_char(byte: u8) -> bool {
  match byte {
    b'a'...b'z' | b'A'...b'Z' => true,
    _ => false
  }
}

fn is_identifier_char(byte: u8) -> bool {
  match byte {
    b'a'...b'z' | b'A'...b'Z' | b'0'...b'9' => true,
    _ => false
  }
}

fn is_digit(byte: u8) -> bool {
  match byte {
    b'0'...b'9' => true,
    _ => false
  }
}

type TokenList<'a> = Vec<Token<'a>>;
type CommentMap<'a> = HashMap<usize, &'a[u8]>;

#[derive(Debug)]
pub enum TokenizeResult<'a> {
  Tokenized(TokenList<'a>, CommentMap<'a>),
  Error(TokenizeError),
}

#[derive(Debug)]
pub enum TokenizeError {
  InvalidBinaryDigitError(usize, usize),
  InvalidHexDigitError(usize, usize),
  InvalidOctalDigitError(usize, usize),
  UnclosedStringError(usize, usize),
  UnclosedInterpolationError(usize, usize),
}

impl<'a> TokenizeResult<'a> {
  pub fn unwrap(self) -> (TokenList<'a>, CommentMap<'a>) {
    match self {
      Tokenized(tokens, comments) => (tokens, comments),
      _ => panic!("Unexpected tokenizer error")
    }
  }
}

impl<'a> Tokenizer<'a> {
  pub fn from_source(source: &'a Vec<u8>) -> Tokenizer<'a> {
    let length = source.len();

    return Tokenizer {
      source: source,
      source_length: length,
    };
  }

  pub fn collect_tokens(&mut self) -> TokenizeResult {
    let source = self.source;
    let length = self.source_length;

    let mut tokens = Vec::new();
    let mut comments = HashMap::new();

    let mut index = 0;
    let mut line = 0;

    let mut string_stack = Vec::new();
    let mut interpolation_stack = Vec::new();

    // We iterate through all chars in a single loop, appending tokens as we find them.
    // The trickiest parts here are related to string interpolations, since they can
    // be nested arbitrarily deep (e.g. "hello $("Ms. $(name)")"). These parts are
    // commented below.
    while index < length {
      let start_index = index;
      let byte = source[start_index];

      if string_stack.is_empty() && byte == b'"' {
        // If the string stack is empty and byte is ", we are at the beginning of
        // a brand new string. Save the start index and advance.
        string_stack.push(index);
        index += 1;
        continue;
      }

      if !string_stack.is_empty() {
        // If the string stack is not empty, we're somewhere inside a string (maybe
        // in an interpolation, though). We must check if we need to end the string,
        // start/end an interpolation, or just carry on.

        if byte == b'"' && string_stack.len() == interpolation_stack.len() {
          // If the two stacks have the same size, we must be inside of an interpolation,
          // so the " indicates the beginning of a nested string literal. Save the index
          // in the string stack and advance.
          string_stack.push(index);
          index += 1;
          continue;
        }

        if byte == b'"' {
          // Here, the " must indicate the end of a string literal section. Pop from
          // the string stack, add a new token, then advance.
          let start_index = string_stack.pop().unwrap();

          tokens.push(StringLiteral {
            start: start_index + 1,
            end: index,
            value: &source[start_index + 1..index],
          });

          index += 1;
          continue;
        }

        if byte == b'$' && start_index + 1 < length && source[start_index + 1] == b'(' {
          // We must be at the beginning of an interpolation, so create a token for
          // the string literal portion leading up to the interpolation, one for the
          // interpolation start, and add to the interpolation stack.
          let string_start_index = string_stack.last().unwrap();

          tokens.push(StringLiteral {
            start: string_start_index + 1,
            end: index,
            value: &source[string_start_index + 1..index],
          });

          tokens.push(InterpolationStart {
            start: start_index,
            end: index + 2,
          });

          interpolation_stack.push(index);
          index += 2;
          continue;
        }

        if byte == b')' {
          // We must be at the end of an interpolation, so make a token for it and
          // fix the index on the last string in the string stack so that it starts
          // here. Decrease the interpolation stack.
          tokens.push(InterpolationEnd {
            start: index,
            end: index + 1,
          });

          string_stack.pop();
          string_stack.push(index);

          interpolation_stack.pop();
          index += 1;
          continue;
        }

        if string_stack.len() > interpolation_stack.len() {
          // If the string stack is larger than the interpolation stack, we must be
          // inside of a string literal portion. Just advance past this char so we can
          // include it in the string later.
          index += 1;
          continue;
        }

        // At this point, we must be inside an interpolation (not a string literal),
        // so continue to collect tokens as we would outside of a string.
      }

      match byte {
        b' ' | b'\r' | b'\t' => {
          index += 1;
        }

        b'\n' => {
          index += 1;
          line += 1;

          tokens.push(LineBreak {
            start: start_index,
            end: index,
          })
        }

        b'(' => {
          index += 1;

          tokens.push(LeftParen {
            start: start_index,
            end: index,
          })
        }

        b')' => {
          index += 1;

          tokens.push(RightParen {
            start: start_index,
            end: index,
          })
        }

        b'{' => {
          index += 1;

          tokens.push(LeftBrace {
            start: start_index,
            end: index,
          })
        }

        b'}' => {
          index += 1;

          tokens.push(RightBrace {
            start: start_index,
            end: index,
          })
        }

        b'[' => {
          index += 1;

          tokens.push(LeftBracket {
            start: start_index,
            end: index,
          })
        }

        b']' => {
          index += 1;

          tokens.push(RightBracket {
            start: start_index,
            end: index,
          })
        }

        b',' => {
          index += 1;

          tokens.push(Comma {
            start: start_index,
            end: index,
          })
        }

        b'.' => {
          index += 1;

          tokens.push(Dot {
            start: start_index,
            end: index,
          })
        }

        b'|' => {
          index += 1;

          tokens.push(Pipe {
            start: start_index,
            end: index,
          })
        }

        b'=' => match source.get(index + 1) {
          Some(b'>') => {
            index += 2;

            tokens.push(DoubleArrow {
              start: start_index,
              end: index,
            })
          }

          _ => {
            index += 1;

            tokens.push(Equals {
              start: start_index,
              end: index,
            })
          }
        },

        b'-' => match source.get(index + 1) {
          Some(b'>') => {
            index += 2;

            tokens.push(Arrow {
              start: start_index,
              end: index,
            })
          }

          _ => {
            index += 1;

            tokens.push(Minus {
              start: start_index,
              end: index,
            })
          }
        },

        b':' => match source.get(index + 1) {
          Some(b'=') => {
            index += 2;

            tokens.push(ColonEquals {
              start: start_index,
              end: index,
            })
          }

          Some(b':') => {
            index += 2;

            tokens.push(DoubleColon {
              start: start_index,
              end: index,
            })
          }

          _ => {
            index += 1;

            tokens.push(Colon {
              start: start_index,
              end: index,
            })
          }
        },

        b'#' => {
          while index < length && source[index] != b'\n' {
            index += 1;
          }

          comments.insert(line, &source[start_index + 1..index]);
        }

        _ if is_identifier_start_char(byte) => {
          while index < length && is_identifier_char(source[index]) {
            index += 1;
          }

          tokens.push(Identifier {
            start: start_index,
            end: index,
            value: &source[start_index..index],
          });

          continue;
        }

        _ if is_digit(byte) => {
          if byte == b'0' {
            match source.get(index + 1) {
              Some(b'b') | Some(b'B') => {
                index += 2;

                while index < length && is_identifier_char(source[index]) {
                  if source[index] != b'0' && source[index] != b'1' {
                    return Error(InvalidBinaryDigitError(index, index + 1))
                  }

                  index += 1;
                }

                tokens.push(BinaryDigits {
                  start: start_index,
                  end: index,
                  value: &source[start_index..index],
                });

                continue;
              },

              Some(b'x') | Some(b'X') => {
                index += 2;

                while index < length && is_identifier_char(source[index]) {
                  if !source[index].is_ascii_hexdigit() {
                    return Error(InvalidHexDigitError(index, index + 1))
                  }

                  index += 1;
                }

                tokens.push(HexDigits {
                  start: start_index,
                  end: index,
                  value: &source[start_index..index],
                });

                continue;
              },

              Some(b'o') | Some(b'O') => {
                index += 2;

                while index < length && is_identifier_char(source[index]) {
                  if source[index] < 48 || source[index] > 55 {
                    return Error(InvalidOctalDigitError(index, index + 1))
                  }

                  index += 1;
                }

                tokens.push(OctalDigits {
                  start: start_index,
                  end: index,
                  value: &source[start_index..index],
                });

                continue;
              },

              _ => {}
            }
          }

          while index < length && is_digit(source[index]) {
            index += 1;
          }

          tokens.push(DecimalDigits {
            start: start_index,
            end: index,
            value: &source[start_index..index],
          });

          continue;
        }

        _ => {
          index += 1;

          tokens.push(Unexpected {
            start: start_index,
            end: index,
          })
        }
      };
    }

    if !interpolation_stack.is_empty() {
      let start_index = interpolation_stack.pop().unwrap();
      return Error(UnclosedInterpolationError(start_index, index))
    }

    if !string_stack.is_empty() {
      let start_index = string_stack.pop().unwrap();
      return Error(UnclosedStringError(start_index, index))
    }

    Tokenized(tokens, comments)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::assert_tokens_snapshot;
  use insta::assert_snapshot;

  assert_tokens_snapshot!(
    empty,
    ""
  );

  assert_tokens_snapshot!(
    identifers,
    "hello world"
  );

  assert_tokens_snapshot!(
    numbers,
    "hello 1 47 wow"
  );

  assert_tokens_snapshot!(
    binary_numbers,
    "0b101 0b00 0b1 0B0 0b00100"
  );

  assert_tokens_snapshot!(
    hex_numbers,
    "0x101 0x0 0xfacade 0X47"
  );

  assert_tokens_snapshot!(
    octal_numbers,
    "0o101 0o0 0o47 0O47"
  );

  assert_tokens_snapshot!(
    comment_before,
    "# comment \nok"
  );

  assert_tokens_snapshot!(
    comment_same_line,
    "ok #comment"
  );

  assert_tokens_snapshot!(
    comment_after,
    "ok \n\n#comment"
  );

  assert_tokens_snapshot!(
    symbols,
    "{ . } ( , ) : [ :: ] := = => ->"
  );

  assert_tokens_snapshot!(
    unexpected,
    "(@$@)"
  );

  assert_tokens_snapshot!(
    strings_without_interpolations,
    "\"hello\" \"\" \"world\""
  );

  assert_tokens_snapshot!(
    strings_with_interpolations,
    "\"hello $(name)!\" nice \"$(str)\""
  );

  assert_tokens_snapshot!(
    strings_with_nested_interpolations,
    "\"hello $(name \"inner $(o)\" wow)!\""
  );
}
