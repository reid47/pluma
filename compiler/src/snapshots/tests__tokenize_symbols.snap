---
source: compiler/src/tokenizer.rs
expression: "{ . } ( , ) : [ :: ] := = => ->"
---
Tokenized(
    [
        LeftBrace {
            start: 0,
            end: 1,
        },
        Dot {
            start: 2,
            end: 3,
        },
        RightBrace {
            start: 4,
            end: 5,
        },
        LeftParen {
            start: 6,
            end: 7,
        },
        Comma {
            start: 8,
            end: 9,
        },
        RightParen {
            start: 10,
            end: 11,
        },
        Colon {
            start: 12,
            end: 13,
        },
        LeftBracket {
            start: 14,
            end: 15,
        },
        DoubleColon {
            start: 16,
            end: 18,
        },
        RightBracket {
            start: 19,
            end: 20,
        },
        ColonEquals {
            start: 21,
            end: 23,
        },
        Equals {
            start: 24,
            end: 25,
        },
        DoubleArrow {
            start: 26,
            end: 28,
        },
        Arrow {
            start: 29,
            end: 31,
        },
    ],
    {},
)
